src/include/migraphx/op/slice.hpp

```diff
@@ -90,80 +190,141 @@ struct slice
+    normalize_inputs(shape input_shape, Inds input_starts, Inds input_ends) const
+    {
+        return {{"input_starts", normalize_indices(input_shape, input_starts, this->axes)},
+                {"input_ends", normalize_indices(input_shape, input_ends, this->axes)}};
+    }
+
+    /**
+     * Three input version of the normalize_inputs.
+     * This one also checks that the input_axes are valid.
+     */
+    template <class Inds>
+    std::unordered_map<std::string, Inds>
+    normalize_inputs(shape input_shape, Inds input_starts, Inds input_ends, Inds input_axes) const
+    {
+        // normalize the axes
+        auto shape_ndim = input_shape.ndim();
+        std::transform(input_axes.begin(), input_axes.end(), input_axes.begin(), [&](auto axis) {
+            if(axis < 0)
             {
-                // TODO: when non-fixed shape slicing is allowed, this will be different than
-                // sliced_length, making use of TBD start/end values.
-                std::size_t sliced_min_length = ends[i] - starts[i];
-                // if the slice size is smaller than maxes but larger than mins
-                new_mins[axis] = std::min(sliced_min_length, new_mins[axis]);
+                axis += shape_ndim;
             }
-        }
-        if(input_shape.dynamic())
+            if(axis < 0 or axis >= shape_ndim)
+            {
+                MIGRAPHX_THROW("SLICE: entry of input_axes out of bounds");
+            }
+            return axis;
+        });
+        return {{"input_starts", normalize_indices(input_shape, input_starts, input_axes)},
+                {"input_ends", normalize_indices(input_shape, input_ends, input_axes)},
+                {"input_axes", input_axes}};
+    }
+
+    argument compute(const shape&, std::vector<argument> args) const
+    {
+        auto input       = args[0];
+        auto input_shape = input.get_shape();
+        std::size_t offset;
+        switch(args.size())
         {
-            return shape{t, new_mins, new_lens, {}};
+        case 1: {
+            offset = compute_offset(input_shape);
+            return {normalize_compute_shape({input_shape}), [=] { return input.data() + offset; }};
+        }
+        case 3: {
+            shape output_shape;
+            visit_all(args[1], args[2])([&](auto input_starts, auto input_ends) {
+                auto norm_inputs = normalize_inputs(input_shape, input_starts, input_ends);
+                offset = compute_offset(input_shape, norm_inputs.at("input_starts"), this->axes);
+                output_shape = {input_shape.type(),
+                                lens_calc(input_shape.lens(),
+                                          norm_inputs.at("input_starts"),
+                                          norm_inputs.at("input_ends"),
+                                          this->axes),
+                                input_shape.strides()};
+            });
+            return {output_shape, [=] { return input.data() + offset; }};
```

Comment:
I think it would be better to remove these returns outside of the switch statement. `output_shape` and `offset` are already defined in the outerscope.

Reply:
What do you mean here? Output shape is not defined in the outer scope

Reply:
I kept the returns because there would have been `break` statements otherwise.

Reply:
Makes sense.